xarray/core/common.py

```diff
@@ -576,7 +576,13 @@ def where(self, cond, other=None, drop=False):
             outobj = self
             outcond = cond
 
-        return outobj._where(outcond)
+        # preserve attributes
+        out = outobj._where(outcond)
+        out.attrs = self.attrs
+        if isinstance(out, Dataset):
```

Comment:
I think it's fine to assume compatibility, given that this is only being called by internal methods. If `other` is not a `Dataset` then something has gone seriously wrong.

On the other hand, I think it might work to call `DataArray.where` with a Dataset as the argument -- worth double checking.